builtin/credential/aws/path_login.go

```diff
@@ -128,13 +128,211 @@ needs to be supplied along with 'identity' parameter.`,
+	case anyIam && !allIam:
+		return logical.ErrorResponse("supplied some of the auth values for the iam auth type but not all"), nil
+	case anyIam:
+		return b.pathLoginResolveRoleIam(ctx, req, data)
+	default:
+		return logical.ErrorResponse("didn't supply required authentication values"), nil
+	}
+}
+
+func (b *backend) pathLoginEc2GetRoleNameAndIdentityDocumentParsed(ctx context.Context, req *logical.Request, data *framework.FieldData) (string, *identityDocument, *logical.Response, error) {
+	identityDocB64 := data.Get("identity").(string)
+	var identityDocBytes []byte
+	var err error
+	if identityDocB64 != "" {
+		identityDocBytes, err = base64.StdEncoding.DecodeString(identityDocB64)
+		if err != nil || len(identityDocBytes) == 0 {
+			return "", nil, logical.ErrorResponse("failed to base64 decode the instance identity document"), nil
+		}
+	}
+
+	signatureB64 := data.Get("signature").(string)
+	var signatureBytes []byte
+	if signatureB64 != "" {
+		signatureBytes, err = base64.StdEncoding.DecodeString(signatureB64)
+		if err != nil {
+			return "", nil, logical.ErrorResponse("failed to base64 decode the SHA256 RSA signature of the instance identity document"), nil
+		}
+	}
+
+	pkcs7B64 := data.Get("pkcs7").(string)
+
+	// Either the pkcs7 signature of the instance identity document, or
+	// the identity document itself along with its SHA256 RSA signature
+	// needs to be provided.
+	if pkcs7B64 == "" && (len(identityDocBytes) == 0 && len(signatureBytes) == 0) {
+		return "", nil, logical.ErrorResponse("either pkcs7 or a tuple containing the instance identity document and its SHA256 RSA signature needs to be provided"), nil
+	} else if pkcs7B64 != "" && (len(identityDocBytes) != 0 && len(signatureBytes) != 0) {
+		return "", nil, logical.ErrorResponse("both pkcs7 and a tuple containing the instance identity document and its SHA256 RSA signature is supplied; provide only one"), nil
+	}
+
+	// Verify the signature of the identity document and unmarshal it
+	var identityDocParsed *identityDocument
+	if pkcs7B64 != "" {
+		identityDocParsed, err = b.parseIdentityDocument(ctx, req.Storage, pkcs7B64)
+		if err != nil {
+			return "", nil, nil, err
+		}
+		if identityDocParsed == nil {
+			return "", nil, logical.ErrorResponse("failed to verify the instance identity document using pkcs7"), nil
+		}
+	} else {
+		identityDocParsed, err = b.verifyInstanceIdentitySignature(ctx, req.Storage, identityDocBytes, signatureBytes)
+		if err != nil {
+			return "", nil, nil, err
+		}
+		if identityDocParsed == nil {
+			return "", nil, logical.ErrorResponse("failed to verify the instance identity document using the SHA256 RSA digest"), nil
+		}
+	}
+
+	roleName := data.Get("role").(string)
+
+	// If roleName is not supplied, a role in the name of the instance's AMI ID will be looked for
+	if roleName == "" {
+		roleName = identityDocParsed.AmiID
+	}
+
+	// Get the entry for the role used by the instance
+	roleEntry, err := b.role(ctx, req.Storage, roleName)
```

Comment:
It seems that these lines exist in two places, is this intentional?

Reply:
To be clear as what I meant by two places, I am referring to `pathLoginUpdateEc2` function. In `pathLoginUpdateEc2`, `pathLoginEc2GetRoleNameAndIdentityDocumentParsed ` is called (which is refactored from the former) and then `roleEntry` is used to check for a CIDR match and other things. 
So, the question is what is getting `roleEntry` used for here? Is it to make sure it exists? If that is the case, would  adding a comment describing the intended use of fetching it be beneficial?

Reply:
Ahh yes I see what you mean. Essentially:

roleEntry does not contain the roleName, but we need to check it to ensure the role exists. It would be possible to return the roleEntry _and_ the roleName to avoid the doubling up but that feels a little strange to me. Ideally, the roleEntry would contain the name and we'd only return and handle that in both functions, but that's not possible.

Reply:
I think adding a comment describing that would be useful!